webapp/channels/src/utils/desktopAPI.ts

```diff
@@ -0,0 +1,288 @@
+        } else {
+            this.postMessageListeners?.set(channel, new Set([listener]));
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    removePostMessageListener = (channel: string, listener: (message: any) => void) => {
+        const set = this.postMessageListeners?.get(channel);
+        set?.delete(listener);
+        if (set?.size) {
+            this.postMessageListeners?.set(channel, set);
+        } else {
+            this.postMessageListeners?.delete(channel);
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    invokeWithMessaging = <T, T2>(
+        sendChannel: string,
+        sendData?: T,
+        receiveChannel?: string,
+    ) => {
+        return new Promise<T2>((resolve) => {
+            /* create and register a temporary listener if necessary */
+            const response = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: T2}}) => {
+                /* ignore messages from other frames */
+                if (origin !== window.location.origin) {
+                    return;
+                }
+
+                /* ignore messages from other channels */
+                if (type !== (receiveChannel ?? sendChannel)) {
+                    return;
+                }
+
+                /* clean up listener and resolve */
+                window.removeEventListener('message', response);
+                resolve(message);
+            };
+
+            window.addEventListener('message', response);
+            window.postMessage(
+                {type: sendChannel, message: sendData},
+                window.location.origin,
+            );
+        });
+    };
+}
+
+const desktopApp = new DesktopApp();
+export default desktopApp;
+
+/*******************************
+ * API Methods
+ *******************************/
+
+/**
+ * Invokes
+ */
+
+export const getBrowserHistoryStatus = () => {
+    if (window.desktopAPI?.requestBrowserHistoryStatus) {
+        return window.desktopAPI.requestBrowserHistoryStatus();
+    }
+
+    return desktopApp.invokeWithMessaging<void, {enableBack: boolean; enableForward: boolean}>(
+        'history-button',
+        undefined,
+        'history-button-return',
+    ).then(({enableBack, enableForward}) => {
+        return {
+            canGoBack: enableBack,
+            canGoForward: enableForward,
+        };
+    });
+};
+
+/**
+ * Listeners
+ */
+
+export const onUserActivityUpdate = (listener: (userIsActive: boolean, idleTime: number, isSystemEvent: boolean) => void) => {
+    if (window.desktopAPI?.onUserActivityUpdate) {
+        window.desktopAPI.onUserActivityUpdate(listener);
+        return () => {};
+    }
+
+    const legacyListener = ({userIsActive, manual}: {userIsActive: boolean; manual: boolean}) => listener(userIsActive, 0, manual);
+    desktopApp.addPostMessageListener('user-activity-update', legacyListener);
+
+    return () => desktopApp.removePostMessageListener('user-activity-update', legacyListener);
```

Comment:
This is where I contradict myself a bit to say that we don't want the rest of the app to be aware of `window.postMessage` :P

Unfortunately we still need to have the app have some control over when to deactivate these listeners, so we do it React-style here by just returning the "off" hook.